### PR TITLE
Add offline tests for azure core streaming 

### PR DESCRIPTION
**Pull request by HimanshuBarak**
# Description

This is in relation to the following issue : https://github.com/Azure/azure-sdk-for-python/issues/9324

**Changes made :**
Added following tests In test_streaming_async.py

- [x] test_decompress_plain_no_header
- [x] test_compress_plain_no_header
- [x] test_decompress_compressed_no_header
- [x] test_compress_compressed_no_header
- [x] test_decompress_plain_header
- [x] test_compress_plain_header
- [x] test_decompress_compressed_header
- [x] test_compress_compressed_header

Added following tests in test_streaming.py

 - [x] test_decompress_plain_no_header
 - [x] test_decompress_compressed_no_header
 - [x] test_compress_plain_header
 -  [x] test_compress_compressed_header
 
 in stream.py added following endpoints support 

-  /plain_header
- /compressed_header

### CODE REVIEW

**Comment by HimanshuBarak** [comment]
@microsoft-github-policy-service agree

ᐧ

**Comment by HimanshuBarak** [comment]
Hey @lmazuel @annatisch  @iscai-msft @xiangyan99 @kashifkhan @johanste @tjprescott  any feedback from your side on the PR? I see the core ci is failing but the issue seems to be due to trio. I saw another issue mentioning this. should i try to fix the trio bug as part of this also ?? please lemme know what should be my next steps.

File: sdk/core/azure-core/tests/async_tests/test_streaming_async.py [hunk 1]
```diff
@@ -180,7 +180,26 @@ async def test_decompress_compressed_header(http_request):
...
+@pytest.mark.asyncio
+@pytest.mark.parametrize("http_request", HTTP_REQUESTS)
+async def test_compress_compressed_no_header_offline(port, http_request):
+    # expect compressed text
+    client = AsyncPipelineClient("")
+    async with client:
+        request = http_request(method="GET", url="http://localhost:{}/streams/compressed_no_header".format(port))
+        pipeline_response = await client._pipeline.run(request,stream=True)
+        response = pipeline_response.http_response
+        data = response.stream_download(client._pipeline,decompress=False)
```
Comment (iscai-msft) [inline sdk/core/azure-core/tests/async_tests/test_streaming_async.py:192]: so to clarify this test, we expect the server to return a body of compressed text. We specifically say "do not decompress", then we want to make sure it fails on decode, right? This makes sense to me, just confirming the intent of the test

File: sdk/core/azure-core/tests/async_tests/test_streaming_async.py [hunk 2]
```diff
@@ -180,7 +180,26 @@ async def test_decompress_compressed_header(http_request):
...
+        content = b""
+        async for d in data:
+            content += d
+        try:
```
Comment (iscai-msft) [inline sdk/core/azure-core/tests/async_tests/test_streaming_async.py:196]: instead, you can use `with pytest.raises()`

```
with pytest.raises("UnicodeDecodeError"):
  content.decode("utf-8")
```

File: sdk/core/azure-core/tests/async_tests/test_streaming_async.py [hunk 3]
```diff
@@ -203,3 +222,126 @@ async def test_compress_compressed_header(http_request):
...
+@pytest.mark.asyncio
+@pytest.mark.parametrize("http_request", HTTP_REQUESTS)
+async def test_decompress_plain_header_offline(port, http_request):
+    # expect error
+    client = AsyncPipelineClient("")
+    async with client:
+        request = http_request(method="GET", url="http://localhost:{}/streams/compressed".format(port))
+        pipeline_response = await client._pipeline.run(request,stream=True)
+        response = pipeline_response.http_response
+        print(f"this is the pipeline respone {response}")
```
Comment (iscai-msft) [inline sdk/core/azure-core/tests/async_tests/test_streaming_async.py:306]: we tend to remove print statements in tests

File: sdk/core/azure-core/tests/async_tests/test_streaming_async.py [hunk 4]
```diff
@@ -203,3 +222,126 @@ async def test_compress_compressed_header(http_request):
...
+        data = response.stream_download(client._pipeline,decompress=True)
+        try:
+            content = b""
+            async for d in data:
+                content += d
+            assert False
+        except (zlib.error, DecodeError):
```
Comment (iscai-msft) [inline sdk/core/azure-core/tests/async_tests/test_streaming_async.py:313]: you can also use `pytest.raises` here, and expect one of these errors to show up

File: sdk/core/azure-core/tests/async_tests/test_streaming_async.py [hunk 5]
```diff
@@ -203,3 +222,126 @@ async def test_compress_compressed_header(http_request):
...
+@pytest.mark.parametrize("http_request", HTTP_REQUESTS)
+async def test_compress_plain_no_header_offline(port, http_request):
+    client = AsyncPipelineClient("")
+    async with client:
+        request = http_request(method="GET", url="http://localhost:{}/streams/string".format(port))
+        pipeline_response = await client._pipeline.run(request,stream=True)
+        response = pipeline_response.http_response
+        data = response.stream_download(client._pipeline,decompress=False)
+        content = b""
+        async for d in data:
```
Comment (iscai-msft) [inline sdk/core/azure-core/tests/async_tests/test_streaming_async.py:327]: don't quote me on this, but I think for the instances when you're async looping through the data, you can instead do `decoded = b''.join([d async for d in data]).decode("utf-8")`, or something like that

**Review by iscai-msft — COMMENTED** [review]
Looks good to me, don't have any suggestions on the content of the test, just some feedback on little things to tweak. Overall looks amazing, thank you so much for contributing!

**Review by kristapratico — COMMENTED** [review]
Thanks for the updates! This looks good to me. We just need to fix the formatting error in CI for core. You can find details on how to do that here: https://github.com/Azure/azure-sdk-for-python/blob/main/doc/eng_sys_checks.md#running-locally

**Review by kristapratico — APPROVED** [review]
Awesome! Thanks a lot for your contribution, @HimanshuBarak! If you're interested, once this merges, you can submit another PR to remove the non-offline tests in test_streaming/_async.py now that we have offline versions.

@iscai-msft - are you good to sign off on this as well?

**Review by iscai-msft — APPROVED** [review]

**Comment by HimanshuBarak** [comment]
Hey @kristapratico I have created the required PR :https://github.com/Azure/azure-sdk-for-python/pull/34247. However I see that the tes_rest_stream_responses_async.py is failing. In local testing this test was skipped with the reason given as " We've gotten rid of iter_text for now". How should I approach this? Is this test still relevant or can I remove this as well?

**Comment by kristapratico** [comment]
@HimanshuBarak I'm re-running the CI pipeline, it looked like one of the envs was failing due to a transient error. Where do you see the error for that particular test? To me, it looks like it should still be skipped.

**Comment by HimanshuBarak** [comment]
@kristapratico   Apologies for this. You are right. It was misinterpretation of logs on my part.  I will be more careful from next time.

**Pull request merged**